lambdaclass/mirra_backend
# [GH-431] Add currency cost to levels

Closes: https://github.com/lambdaclass/champions_of_mirra/issues/431

Associated PR: https://github.com/lambdaclass/champions_of_mirra/issues/463

## Motivation

We need to be able to charge the user with a currency cost in order to get access to fight some of the levels. For example, the Dungeon levels will have a cost in Shovels.

## Summary of changes
- Add an embedded array of CurrencyCost to the Level schema.
- Before starting a battle, check that the user can afford it.
- Substract the level's cost in currency when fighting the battle.

## How to test it?

Run the backend and start the client in the main branch. For starters, everything should be working as usual, when none of the existing levels have a CurrencyCost associated.

Add this line to the map used for level creation in seeds.ex: `currency_costs: [%CurrencyCost{currency_id: gold_currency.id, amount: 13}],`
Also, you'll need to add this alias to that file: `alias GameBackend.Users.Currencies.CurrencyCost`

You should end up with something like this:

```elixir
levels =
  Enum.flat_map(Enum.with_index(rules, 1), fn {campaign_rules, campaign_index} ->
    {_, campaign} =
      Campaigns.insert_campaign(
        %{
          game_id: champions_of_mirra_id,
          super_campaign_id: super_campaign.id,
          campaign_number: campaign_index
        },
        returning: true
      )

    Enum.map(1..campaign_rules.length, fn level_index ->
      %{
        game_id: champions_of_mirra_id,
        campaign_id: campaign.id,
        level_number: level_index,
        experience_reward: 100 * level_index,
        attempt_cost: [%CurrencyCost{currency_id: gold_currency.id, amount: 13}],
        inserted_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second),
        updated_at: NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
      }
    end)
  end)
```

Now all the levels cost 13 Gold. Before fighting a new level, check your current amount of gold. If you fight the level and lose the battle, your amount of gold should have decreased by 13. If you won, your gold amount should now be `previous amount + level reward - level cost`.

## Checklist
